extensions.html

```diff
@@ -209,33 +338,356 @@ <h2>
+                </li>
+                <li>Set
+                [=this=].{{GamepadHapticActuator/[[playingEffectPromise]]}} to
+                `undefined`.
+                </li>
+              </ol>
+            </li>
+            <li>[=Stop haptic effects=] on this actuator.
+            </li>
+            <li>Return [=a promise resolved with=]
+            {{GamepadHapticsResult/"complete"}};
+            </li>
+          </ol>
+        </dd>
+        <dt>
+          Construction
+        </dt>
+        <dd>
+          <p>
+            Add new steps to [=a new `Gamepad`=]:
+          </p>
+          <ol>
+            <li>If |gamepad:Gamepad| has a vibration actuator that can [=play
+            effects with type=] {{GamepadHapticEffectType/"dual-rumble"}},
+            initialize |gamepad:Gamepad|'s {{Gamepad/vibrationActuator}}
+            attribute to [=a new `GamepadHapticActuator`=] with
+            {{GamepadHapticActuatorType/"dual-rumble"}}.
+            </li>
+            <li>Otherwise, initialize |gamepad:Gamepad|'s
+            {{Gamepad/vibrationActuator}} attribute to `null`.
+            </li>
+          </ol>
+          <p>
+            <dfn>A new `GamepadHapticActuator`</dfn> with
+            |type:GamepadHapticEffectType| is constructed by performing the
+            following steps:
+          </p>
+          <ol>
+            <li>Let |actuator:GamepadHapticActuator| be a newly created
+            {{Gamepad}} instance:
+              <ol>
+                <li>Initialize |actuator|'s {{GamepadHapticActuator/type}}
+                attribute to |type|.
+                </li>
+              </ol>
+            </li>
+            <li>Return |actuator|.
+            </li>
+          </ol>
+        </dd>
+        <dt>
+          Page visibility
+        </dt>
+        <dd>
+          <p>
+            This specification defines an [=external now hidden algorithm=].
+            When the |document|'s [=visibility state=] becomes `"hidden"`, run
+            these steps for each {{GamepadHapticActuator}}
+            |actuator:GamepadHapticActuator|:
+          </p>
+          <ol>
+            <li>If
+            |actuator|.{{GamepadHapticActuator/[[playingEffectPromise]]}} is
+            `undefined`, abort these steps.
+            </li>
+            <li>[=Queue a global task=] on the [=relevant global object=] of
+            |actuator| using the [=gamepad task source=] to run the following
+            steps:
+              <ol>
+                <li>[=Resolve=]
+                |actuator|.{{GamepadHapticActuator/[[playingEffectPromise]]}}
+                with {{GamepadHapticsResult/"preempted"}}.
+                </li>
+                <li>Set
+                |actuator|.{{GamepadHapticActuator/[[playingEffectPromise]]}}
+                to `undefined`.
+                </li>
+              </ol>
+            </li>
+            <li>[=Stop haptic effects=] on |actuator|.
```

Comment:
See above... this should probably first stop the effects, get IPC confirmation, and then resolve `playingEffectPromise` with "preempted". Otherwise, this will be racy (stopping the effects might take longer than resolving the promise).

Reply:
I'm not sure if this is possible. Typically we don't get any feedback from the gamepad when an actuator is started or stopped. The best we can do is confirm that the command was sent to the gamepad.